pkg/controller/job/job_controller.go

```diff
@@ -1554,13 +1583,18 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
 	return calculated
 }
 
-// countValidPodsWithFilter returns number of valid pods that pass the filter.
+// getValidPodsWithFilter returns the valid pods that pass the filter.
 // Pods are valid if they have a finalizer and, for Indexed Jobs, a valid
 // completion index.
-func countValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) int {
-	result := len(uncounted)
+func getValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) []*v1.Pod {
+	result := []*v1.Pod{}
 	for _, p := range pods {
 		uid := string(p.UID)
+
+		if uncounted.Has(uid) {
```

Comment:
It this needed? Note that, it might add to the `result` pods which are not accepted by the filter. It would be good to use the same logic, just append instead of incrementing the counter.

Reply:
In the master, the result is initialized to `len(uncounted)` as in this [permalink](https://github.com/kubernetes/kubernetes/blob/e55f2a9b54e0f6bd9ed12e50752d7d2545ab0cab/pkg/controller/job/job_controller.go#L1561); not with zero. This new `if` condition that I introduced does exactly the same. Let me know if this makes sense.

Reply:
Oh, I see, this is fine. Please resolve.